### ISSUE TITLE
Incorrect output from MATCH with "class" and "while" clauses

### ISSUE THREAD

**Issue by obi1kenobi** [orientechnologies/orientdb#7823]
## Expected behavior  
Regardless of other `while` and `where` clauses, a MATCH statement's `class: Bar` clause should behave identically to a `where: (@this INSTANCEOF 'Bar')`.

## Actual behavior  
In the query below, using the `class:` clause causes the query to incorrectly return no results, while using the equivalent `where:` clause returns the correct set of results.

## Steps to reproduce  
Bear with me, this is a hard one to reproduce. This is the minimal setup I've been able to consistently reproduce the issue with:
```
CREATE CLASS Foo EXTENDS V
CREATE CLASS AbstractBar EXTENDS V
CREATE CLASS Bar EXTENDS AbstractBar
CREATE CLASS Baz EXTENDS V
CREATE PROPERTY Foo.name STRING
CREATE PROPERTY AbstractBar.name STRING
CREATE PROPERTY Baz.name STRING

CREATE CLASS Foo_Bar EXTENDS E
CREATE CLASS Bar_ParentBar EXTENDS E
CREATE CLASS Baz_Foo EXTENDS E

# tier 2:           bar-1-tier-2
#                    /        \
# tier 1:  bar-1-tier-1     bar-2-tier-1
CREATE VERTEX Bar SET name = 'bar-1-tier-1'
CREATE VERTEX Bar SET name = 'bar-2-tier-1'
CREATE VERTEX Bar SET name = 'bar-1-tier-2'

CREATE EDGE Bar_ParentBar FROM (SELECT * FROM Bar WHERE name != 'bar-1-tier-2') TO (SELECT * FROM Bar WHERE name = 'bar-1-tier-2')

CREATE VERTEX Foo SET name = 'foobaz1'
CREATE VERTEX Foo SET name = 'foobaz2'
CREATE VERTEX Foo SET name = 'foobaz3'
CREATE VERTEX Foo SET name = 'foobaz4'
CREATE VERTEX Foo SET name = 'foobaz5'

# Foos 1-3 are connected to the 'bar-1-tier-1' Bar
# Foos 4-5 are connected to the 'bar-2-tier-1' Bar
CREATE EDGE Foo_Bar FROM (SELECT * FROM Foo WHERE name <= 'foobaz3') TO (SELECT * FROM Bar WHERE name = 'bar-1-tier-1')
CREATE EDGE Foo_Bar FROM (SELECT * FROM Foo WHERE name > 'foobaz3') TO (SELECT * FROM Bar WHERE name = 'bar-2-tier-1')

CREATE VERTEX Baz SET name = 'foobaz1'
CREATE VERTEX Baz SET name = 'foobaz2'
CREATE VERTEX Baz SET name = 'foobaz3'
CREATE VERTEX Baz SET name = 'foobaz4'
CREATE VERTEX Baz SET name = 'foobaz5'

# Each Baz is connected to the Foo with the matching name
CREATE EDGE Baz_Foo FROM (SELECT * FROM Baz WHERE name = 'foobaz1') TO (SELECT * FROM Foo WHERE name = 'foobaz1')
CREATE EDGE Baz_Foo FROM (SELECT * FROM Baz WHERE name = 'foobaz2') TO (SELECT * FROM Foo WHERE name = 'foobaz2')
CREATE EDGE Baz_Foo FROM (SELECT * FROM Baz WHERE name = 'foobaz3') TO (SELECT * FROM Foo WHERE name = 'foobaz3')
CREATE EDGE Baz_Foo FROM (SELECT * FROM Baz WHERE name = 'foobaz4') TO (SELECT * FROM Foo WHERE name = 'foobaz4')
CREATE EDGE Baz_Foo FROM (SELECT * FROM Baz WHERE name = 'foobaz5') TO (SELECT * FROM Foo WHERE name = 'foobaz5')
```

Here's the query with `where:` and `INSTANCEOF`:
```
MATCH {
    class: Foo,
    as: my_starting_foo
}.out('Foo_Bar') {
    where: (@this INSTANCEOF 'Bar'),
    as: my_starting_bar
}.in('Bar_ParentBar') {
    while: ($depth < 10),
    as: my_recursing_bar
}.in('Foo_Bar') {
    class: Foo,
    where: ((name <> $matched.my_starting_foo.name)),
    as: my_final_foo
}.in('Baz_Foo') {
    class: Baz,
    as: my_baz
} RETURN $matches
```
It correctly outputs 8 results: the entire block with `while:` is a no-op, and the valid results are any pair of the 3 `Foo`s connected to the same tier-1 `Bar`, and the two permutations of the 2 `Foo`s connected to the same tier-1 `Bar` on the other side.

Let's make that `INSTANCEOF` into a `class:` clause:
```
MATCH {
    class: Foo,
    as: my_starting_foo
}.out('Foo_Bar') {
    class: Bar,
    as: my_starting_bar
}.in('Bar_ParentBar') {
    while: ($depth < 10),
    as: my_recursing_bar
}.in('Foo_Bar') {
    class: Foo,
    where: ((name <> $matched.my_starting_foo.name)),
    as: my_final_foo
}.in('Baz_Foo') {
    class: Baz,
    as: my_baz
} RETURN $matches
```
Now, the query returns no results. This is clearly incorrect.

Hope you are able to make sense of this -- please let me know if I can help with anything.

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

It's definitely a bug, I'm checking it

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

I just pushed a fix on 2.2.x branch, it will be released with 2.2.30.

Now let's see how to port it to 3.0

Thanks

Luigi

**Comment by obi1kenobi** [comment]
Fantastic, thank you for the speedy resolution. Looking forward to 2.2.30!
